src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,698 @@
+    return _listdir_from_keys(store, listdir_path)
+
+
+class LRUStoreCache(Store):
+    """
+    Storage class that implements a least-recently-used (LRU) cache layer over
+    some other store.
+
+    Intended primarily for use with stores that can be slow to
+    access, e.g., remote stores that require network communication to store and
+    retrieve data.
+
+    The cache stores the raw bytes returned by the underlying store, before any
+    decompression or array processing. This means that compressed data remains
+    compressed in the cache, and decompression happens each time the cached data
+    is accessed. This design choice keeps the cache lightweight while still
+    providing significant performance benefits for network-bound operations.
+
+    This store supports both read and write operations. Write operations use a
+    write-through strategy where data is written to both the underlying store
+    and cached locally. The cache automatically invalidates entries when the
+    underlying data is modified to maintain consistency.
+
+    Parameters
+    ----------
+    store : Store
+        The store containing the actual data to be cached.
+    max_size : int
+        The maximum size that the cache may grow to, in number of bytes.
+        This parameter is required to prevent unbounded memory growth.
+
+        Values smaller than your typical chunk size will result in most data
+        being silently excluded from the cache, reducing effectiveness.
+
+    Examples
+    --------
+    The example below wraps a LocalStore with an LRU cache for demonstration::
+
+        >>> import tempfile
+        >>> import zarr
+        >>> from zarr.storage import LocalStore
+        >>>
+        >>> # Create a temporary directory for the example
+        >>> temp_dir = tempfile.mkdtemp()
+        >>> store = LocalStore(temp_dir)
+        >>>
+        >>> # Create some test data first
+        >>> arr = zarr.create((1000, 1000), chunks=(100, 100), store=store, dtype='f4')
+        >>> arr[:] = 42.0
+        >>>
+        >>> # Now wrap with cache for faster access
+        >>> cached_store = zarr.LRUStoreCache(store, max_size=1024 * 1024 * 256)  # 256MB cache
+        >>> cached_arr = zarr.open(cached_store)
+        >>>
+        >>> # First access loads from disk and caches
+        >>> data1 = cached_arr[0:100, 0:100]  # Cache miss
+        >>>
+        >>> # Second access uses cache (much faster for remote stores)
+        >>> data2 = cached_arr[0:100, 0:100]  # Cache hit
+
+    For remote stores where the performance benefit is more apparent::
+
+        >>> from zarr.storage import RemoteStore
+        >>> # remote_store = RemoteStore.from_url("https://example.com/data.zarr")
+        >>> # cached_remote = zarr.LRUStoreCache(remote_store, max_size=2**28)
+
+
+    """
+
+    supports_writes: bool = True
+    supports_deletes: bool = True
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
```

Comment:
I think this will not always be `True`, because it depends on if the underlying store supports listing. In fact, I think *all* these properties will depend on the underlying store. So they should be re-implemented as properties that return the values from the underlying store.